File: programs/libc/src/unistd/reboot.c

#include "user_syscall.h"

extern int errno;

/*
 * Reboot the computer
 */
int reboot(void)
{
	int ret = _user_syscall(REBOOT, 0);
	/*
	 * On Error, -1 is returned, and errno is set appropriately.
	 */
	if (ret < 0) {
		errno = -ret;
		return -1;
	} else {
		errno = 0;
		return ret;
	}
}

File: rust_kernel/src/taskmaster/syscall/reboot.rs
//! sys_reboot and sys_shutdown implementations

use super::{SysResult, SCHEDULER};

use libc_binding::Errno;

use crate::drivers::ACPI;

/// Reboot thw computer
pub fn sys_reboot() -> SysResult<u32> {
    unpreemptible_context!({
        match *ACPI.lock() {
            Some(mut acpi) => match acpi.reboot_computer() {
                Ok(_) => {}
                Err(e) => {
                    log::error!(
                        "ACPI reboot failure: {:?}. Trying with PS/2 controler ...",
                        e
                    );
                    // Try to reboot with PS/2 controler
                    SCHEDULER.lock().reboot_computer();
                }
            },
            None => {
                // Try to reboot with PS/2 controler
                SCHEDULER.lock().reboot_computer();
            }
        }
    });
    Err(Errno::EACCES)
}
